src/mdl/holder.rs

```diff
@@ -0,0 +1,536 @@
+        Some(drms),
+        None,
+    )
+    .map_err(|e| SessionError::Generic {
+        value: format!("Could not initialize session: {e:?}"),
+    })?;
+    let ble_ident = session
+        .ble_ident()
+        .map_err(|e| SessionError::Generic {
+            value: format!("Couldn't get BLE identification: {e:?}").to_string(),
+        })?
+        .to_vec();
+    let (engaged_state, qr_code_uri) =
+        session.qr_engagement().map_err(|e| SessionError::Generic {
+            value: format!("Could not generate qr engagement: {e:?}"),
+        })?;
+    Ok(MdlPresentationSession {
+        engaged: Mutex::new(Some(engaged_state)),
+        in_process: Mutex::new(None),
+        qr_code_uri,
+        ble_ident,
+    })
+}
+
+#[derive(uniffi::Object)]
+pub struct MdlPresentationSession {
+    engaged: Mutex<Option<device::SessionManagerEngaged>>,
+    in_process: Mutex<Option<InProcessRecord>>,
+    pub qr_code_uri: String,
+    pub ble_ident: Vec<u8>,
+}
+
+#[derive(uniffi::Object, Clone)]
+struct InProcessRecord {
+    session: device::SessionManager,
+    items_request: device::RequestedItems,
+}
+
+#[uniffi::export]
+impl MdlPresentationSession {
+    /// Handle a request from a reader that is seeking information from the mDL holder.
+    ///
+    /// Takes the raw bytes received from the reader by the holder over the transmission
+    /// technology. Returns a Vector of information items requested by the reader, or an
+    /// error.
+    pub fn handle_request(&self, request: Vec<u8>) -> Result<Vec<ItemsRequest>, RequestError> {
+        // Mutexes only return Err if another thread has panicked while holding the mutex
+        // If that has happened, its probably better to just crash. This is what the standard documentation recommends.
+        // See https://doc.rust-lang.org/std/sync/struct.Mutex.html
+        let (session_manager, items_requests) = {
+            let session_establishment: SessionEstablishment = serde_cbor::from_slice(&request)
+                .map_err(|e| RequestError::Generic {
+                    value: format!("Could not deserialize request: {e:?}"),
+                })?;
+            let engaged = self
+                .engaged
+                .lock()
+                .expect("Could not lock mutex")
+                .deref_mut()
+                .take();
+            engaged
+                .unwrap()
+                .process_session_establishment(session_establishment)
+                .map_err(|e| RequestError::Generic {
+                    value: format!("Could not process process session establishment: {e:?}"),
+                })?
+        };
+
+        let mut in_process = self.in_process.lock().unwrap();
+        *in_process = Some(InProcessRecord {
+            session: session_manager,
+            items_request: items_requests.clone(),
+        });
+        Ok(in_process
+            .as_ref()
+            .unwrap()
```

Comment:
Can you return an error instead of panicking?

Reply:
This unwrap can't panic because we set `in_process` to `Some()` in the previous line. I've added a comment to that effect.

Reply:
I've realized that I don't need to do this at all. `items_request` is still in scope at this point, so I can use it directly, avoiding the unwrap entirely.